File: src/cpp/FindPath.cpp
/*
struct TreeNode {
	int val;
	struct TreeNode *left;
	struct TreeNode *right;
	TreeNode(int x) :
			val(x), left(NULL), right(NULL) {
	}
};*/
class Solution {
public:



    vector<vector<int> > FindPath(TreeNode* root,int expectNumber) {
		if (root == NULL) {
            return vector<vector<int> >();
        }

        int currSum = 0;
        vector<int> path;
        vector<vector<int> > result;

        FindPaths(root, expectNumber, path, result, currSum);

        return result;

    }

    void FindPaths(TreeNode* root, int expectNumber, vector<int> path, vector<vector<int> >& result, int currSum) {

        currSum += root->val;
        path.push_back(root->val);
        bool isLeaf = (root->left == NULL) && (root->right == NULL);

        if (currSum == expectNumber && isLeaf) {
            result.push_back(path);
        }

        if (root->left != NULL) {
            FindPaths(root->left, expectNumber, path, result, currSum);
        }

        if (root->right != NULL) {
            FindPaths(root->right, expectNumber, path, result, currSum);
        }

        path.pop_back();

    }

};

File: src/java/FindPath.java
import java.util.ArrayList;
import java.util.Stack;
/**
public class TreeNode {
    int val = 0;
    TreeNode left = null;
    TreeNode right = null;

    public TreeNode(int val) {
        this.val = val;

    }

}
*/
public class Solution {
    public ArrayList<ArrayList<Integer>> FindPath(TreeNode root,int target) {
        ArrayList<ArrayList<Integer>> pathList = new ArrayList<ArrayList<Integer>>();
        if (root == null) {
            return pathList;
        }
        Stack<Integer> stack = new Stack<>();
        FindPath(root, target, stack, pathList);
        return pathList;
    }

    private void FindPath(TreeNode root, int target, Stack<Integer> path, ArrayList<ArrayList<Integer>> pathList) {
        if (root == null) {
            return;
        }
        if (root.left == null && root.right == null) {
            if (root.val == target) {
                ArrayList<Integer> list = new ArrayList<>();
                for (int i : path) {
                    list.add(new Integer(i));
                }
                list.add(new Integer(root.val));
                pathList.add(list);
            }
        } else {
            path.push(new Integer(root.val));
            if (root.left != null) {
                FindPath(root.left, target - root.val, path, pathList);
            }
            if (root.right != null) {
                FindPath(root.right, target - root.val, path, pathList);
            }
            path.pop();
        }
    }
}
